File: android/src/main/kotlin/com/jojo/action_call_state/handler/ActionCallStateHandler.kt
package com.jojo.action_call_state.handler

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.telephony.TelephonyManager
import androidx.annotation.NonNull
import com.jojo.action_call_state.receiver.ActionCallStateBroadcast
import com.jojo.action_call_state.utils.Constants
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.EventSink
import org.json.JSONObject

@Suppress("DEPRECATION")
class ActionCallStateHandler(@NonNull binding: FlutterPluginBinding){
    private var actionCallEventChannel: EventChannel =
        EventChannel(binding.binaryMessenger, Constants.EVENT_CHANNEL)

    init {
        actionCallEventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            private var systemType: String? = null
            private lateinit var receiver: ActionCallStateBroadcast
            override fun onListen(arguments: Any?, events: EventSink?) {
                val intentFilterButton = IntentFilter()
                intentFilterButton.addAction("com.yunke.ACTION_INCOMING_CALL")
                intentFilterButton.addAction("com.yunke.ACTION_OUTGOING_CALL")
                intentFilterButton.addAction("com.yunke.ACTION_CALL_DISCONNECTED")
                intentFilterButton.addAction("com.yunke.ACTION_CALL_REJECT")
                intentFilterButton.addAction("com.yunke.ACTION_CALL_CONNECTED")
                intentFilterButton.addAction(Intent.ACTION_NEW_OUTGOING_CALL)
                intentFilterButton.addAction(TelephonyManager.ACTION_PHONE_STATE_CHANGED)
                receiver = object: ActionCallStateBroadcast() {
                    override fun onReceive(context: Context?, intent: Intent?) {
                        super.onReceive(context, intent)
                        if (systemMark == Constants.SYSTEN_MARK_YUNKE) {
                            systemType = Constants.SYSTEN_MARK_YUNKE
                        }
                        events?.success(
                            JSONObject(
                                mapOf(
                                    "systemMark" to (systemType ?: systemMark),
                                    "callStatus" to actionCallStatus.ordinal,
                                    "phoneNumber" to phoneNumber,
                                    "slotIndex" to slotIndex
                                )
                            ).toString()
                        )
                    }
                }

                binding.applicationContext.registerReceiver(
                    receiver,
                    intentFilterButton
                )
            }

            override fun onCancel(arguments: Any?) {
                binding.applicationContext.unregisterReceiver(receiver)
            }

        })
    }

    fun dispose () {
        actionCallEventChannel.setStreamHandler(null)
    }

}

File: lib/src/handler/action_call_state_handler.dart
import 'dart:convert';

import 'package:action_call_state/src/data/action_call_data.dart';
import 'package:action_call_state/src/interface/action_call_interface.dart';
import 'package:action_call_state/src/utils/action_call_status.dart';
import 'package:action_call_state/src/utils/constants.dart';
import 'package:action_call_state/src/utils/system_mark_type.dart';
import 'package:flutter/services.dart';

class ActionCallStateHandler implements ActionCallInterface {
  final actionCallEventChannel = const EventChannel(Constants.eventChannel);

  @override
  Stream<ActionCallData> stream() {
    return actionCallEventChannel.receiveBroadcastStream().distinct().map((event) {
      Map<String, dynamic> map = jsonDecode(event);
      return ActionCallData(
        callState: ActionCallStatus.fromInt(map["callStatus"]),
        systemMarkType: SystemMarkType.fromString(map["systemMark"]),
        phoneNumber: map["phoneNumber"],
        slotIndex: map["slotIndex"],
      );
    });
  }
}
